contracts/stake-types/src/lib.rs

```diff
@@ -0,0 +1,91 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+//! Types used for transactions with Dusk's stake contract.
+
+#![no_std]
+#![deny(missing_docs)]
+#![deny(clippy::pedantic)]
+
+extern crate alloc;
+use alloc::vec::Vec;
+
+mod sig;
+mod stake;
+
+pub use sig::*;
+pub use stake::*;
+
+use dusk_bls12_381_sign::{PublicKey, Signature};
+use dusk_pki::StealthAddress;
+use dusk_plonk::prelude::*;
+use phoenix_core::Note;
+
+use bytecheck::CheckBytes;
+use rkyv::{Archive, Deserialize, Serialize};
+
+/// Stake a value on the stake contract.
+#[derive(Debug, Clone, Archive, Deserialize, Serialize)]
+#[archive_attr(derive(CheckBytes))]
+pub struct Stake {
+    /// Public key to which the stake will belong.
+    pub public_key: PublicKey,
+    /// Signature belonging to the given public key.
+    pub signature: Signature,
+    /// Value to stake.
+    pub value: u64,
+    /// Proof of the `STCT` circuit.
+    pub proof: Proof,
+}
+impl Stake {
+    /// Serializes stake in a vector of bytes
+    #[must_use]
+    pub fn serialize(&self) -> Vec<u8> {
+        rkyv::to_bytes::<_, 8192>(self)
+            .expect("Serializing stake transaction should succeed")
+            .to_vec()
+    }
+}
+
+/// Unstake a value from the stake contract.
+#[derive(Debug, Clone, Archive, Deserialize, Serialize)]
+#[archive_attr(derive(CheckBytes))]
+pub struct Unstake {
+    /// Public key to unstake.
+    pub public_key: PublicKey,
+    /// Signature belonging to the given public key.
+    pub signature: Signature,
+    /// Note to withdraw to.
+    pub note: Note,
+    /// A proof of the `WFCT` circuit.
+    pub proof: Proof,
+}
+
+/// Withdraw the accumulated reward.
+#[derive(Debug, Clone, Archive, Deserialize, Serialize)]
+#[archive_attr(derive(CheckBytes))]
+pub struct Withdraw {
+    /// Public key to withdraw the rewards.
+    pub public_key: PublicKey,
+    /// Signature belonging to the given public key.
+    pub signature: Signature,
+    /// The address to mint to.
+    pub address: StealthAddress,
+    /// A nonce to prevent replay.
+    pub nonce: BlsScalar,
```

Comment:
Is this required? Iiuc the signature already include the counter to prevent replays

Reply:
The counter does prevent replay inside the stake contract, you're right. However, the transfer contract still requires a nonce to mint a note.